### PR TITLE
Rewrite ChainerX indexing tests 

### PR DESCRIPTION
**Pull request by niboshi**
Part of #6401

~Depends on #6437~
~Depends on  #6526~

### CODE REVIEW

**Comment by niboshi** [comment]
Please review #6526 first.

File: tests/chainerx_tests/op_utils.py [hunk 1]
```diff
@@ -291,15 +401,12 @@ def wrap(cls):
...
-        ]
         for cls, mod in classes:
-            for func_suffix, method_name in tests:
-                _create_test_entry_function(
-                    cls, sys.modules[mod], devices, func_suffix, method_name)
+            if not issubclass(cls, OpTest):
+                raise TypeError(
+                    '@op_test decorator can only be applied to OpTest class '
+                    'definition.')
+            _create_test_entry_function(cls, sys.modules[mod], devices)
```
Comment (hvy) [inline tests/chainerx_tests/op_utils.py]: Is this function name still appropriate? The logic is changed quite a bit.

File: tests/chainerx_tests/op_utils.py [hunk 2]
```diff
@@ -174,24 +174,111 @@ def forward_xp(self, inputs, xp):
...
 
+    # Acceptable errors in forward computation.
+    forward_accept_errors = ()
+
+    # Detected acceptable error in forward_chainerx / forward_expected.
+    # None     : Not computed yet
+    # 'ok'     : Computed without error.
+    # Exception: Error was detected.
+    __forward_error_chainerx = None
+    __forward_error_expected = None
```
[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainerx_tests/op_utils.py]: Not needed, but it's just a better practice.

File: tests/chainerx_tests/op_utils.py (hunk 1 above)
Comment (niboshi) [reply tests/chainerx_tests/op_utils.py]: Yes, I think so.

File: tests/chainerx_tests/op_utils.py [hunk 3]
```diff
@@ -174,24 +174,111 @@ def forward_xp(self, inputs, xp):
...
+        accept_errors = self.__get_accept_errors()
+        try:
+            outputs = self.forward_xp(inputs, numpy)
+            self.__forward_error_expected = 'ok'
+        except accept_errors as e:
+            # Keep detected error
+            self.__forward_error_expected = e
+            # A dummy output array is returned
+            y = numpy.zeros((0,), 'float32')
+            outputs = y,
```
[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainerx_tests/op_utils.py]: Do you mean `chainerx.testing.ignore()`?
`FunctionTestBase` does not recognize it.

File: tests/chainerx_tests/unit_tests/routines_tests/test_indexing.py [hunk 4]
```diff
@@ -101,15 +103,24 @@
...
 ])
-def test_getitem(xp, shape, indices):
-    a = array_utils.create_dummy_ndarray(xp, shape, 'int32')
-    return a[indices]
+class TestGetitem(op_utils.NumpyOpTest):
+    # TODO(niboshi): Remove this
+    check_numpy_strides_compliance = False
+
+    def generate_inputs(self):
+        x = array_utils.create_dummy_ndarray(numpy, self.shape, 'float32')
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_indexing.py]: How about getting rid of the `array_utils` dependency in this file by generating inputs similar to `TestTake` below? You should be able to use the `contiguous` attribute for testing "padded" arrays.

File: tests/chainerx_tests/op_utils.py [hunk 5]
```diff
@@ -217,8 +304,7 @@ def _make_backend_config(device_name):
     return backend_config
 
 
-def _create_test_entry_function(
-        cls, module, devices, func_suffix, method_name):
+def _create_test_entry_function(cls, module, devices):
```
Comment (hvy) [inline tests/chainerx_tests/op_utils.py]: Could you please update the docs for this function, e.g. the removal of `func_suffix`?

File: tests/chainerx_tests/unit_tests/routines_tests/test_indexing.py (hunk 4 above)
Comment (niboshi) [reply tests/chainerx_tests/unit_tests/routines_tests/test_indexing.py]: Fixed.

>You should be able to use the contiguous attribute for testing "padded" arrays.

Input arrays are non-contiguous by default.

File: tests/chainerx_tests/op_utils.py (hunk 5 above)
Comment (niboshi) [reply tests/chainerx_tests/op_utils.py]: Fixed.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by niboshi** [comment]
Rebased and resolved conflict.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
CI failures look irrelevant. LGTM!

**Pull request merged**